dora/core/server/worker/src/main/java/alluxio/worker/netty/PacketReadTaskStateMachine.java

```diff
@@ -488,6 +491,9 @@ public void operationComplete(ChannelFuture future) throws InterruptedException
       if (!future.isSuccess()) {
         LOG.error("Failed to send packet.", future.cause());
         mFlowControlQueue.take();
+        if (mAllDataRead && mFlowControlQueue.isEmpty()) {
+          mWaitForTerminatedSignalQueue.offer("wait for finishing sending data");
```

Comment:
hmm if you don't send this message then the getter waits indefinitely?

Reply:
This method is called asyncronously when the a packet is sent successfully. But there may be some packets that have not been sent yet. So we check if all the packets are sent here and put a signal into the queue `mWaitForTerminatedSignalQueue` to notify the main thread that we have finishing sending in `onTerminatedNormally`.